README.md
# WP Toolbelt

[![Travis build process](https://travis-ci.com/BinaryMoon/wp-toolbelt.svg?branch=master)](https://travis-ci.com/github/BinaryMoon/wp-toolbelt)

A lightweight plugin inspired by Jetpack but designed for privacy, speed, and accessibility.

## Speed?

I want it to be fast. I want to be able to enable any of the features and not have my Google Pagespeed score drop. To be fast it:

* Doesn't use jQuery or any other javascript framework. All javascript is vanilla js, and minified.
* Minifies all assets (JS and CSS)
* Loads all assets inline. They are already small, and loading them directly on the page means there are no server requests.
* Only loads things when they are needed. JS and CSS are only loaded for activated modules.
* Options? There's only one for the database; an array that stores what modules are active.
* Uses the minimum code possible. Minimum Javascript and PHP. Less code means more speed, and fewer bugs.
* All modules are disabled by default. You only enable the ones you need.

## Privacy? GDPR?

To ensure the plugin is as privacy focused as possible it:

* Does not phone out. No data is shared with third parties.
* Does not use standard social sharing javascripts (loaded from social networks servers).
* Does not track your usage of the plugin.
* Does not add generator comments, or secret comments to your site html.

## Features

* [Admin Tweaks](https://github.com/BinaryMoon/wp-toolbelt/wiki/Admin-Tweaks)
* [Breadcrumbs](https://github.com/BinaryMoon/wp-toolbelt/wiki/Breadcrumbs)
* [Contact Form](https://github.com/BinaryMoon/wp-toolbelt/wiki/Contact-Form)
* [Cookie Banner](https://github.com/BinaryMoon/wp-toolbelt/wiki/Cookie-Banner)
* [CSP Header](https://github.com/BinaryMoon/wp-toolbelt/wiki/CSP-Header)
* [Disable Comment Urls](https://github.com/BinaryMoon/wp-toolbelt/wiki/Disable-Comment-Urls)
* [Fast 404](https://github.com/BinaryMoon/wp-toolbelt/wiki/Fast-404)
* [Featured Attachments](https://github.com/BinaryMoon/wp-toolbelt/wiki/Featured-Attachment)
* [Footnotes](https://github.com/BinaryMoon/wp-toolbelt/wiki/Footnotes)
* [Heading Anchors](https://github.com/BinaryMoon/wp-toolbelt/wiki/Heading-Anchor)
* [Infinite Scroll](https://github.com/BinaryMoon/wp-toolbelt/wiki/Infinite-Scroll)
* [Lazy Loading](https://github.com/BinaryMoon/wp-toolbelt/wiki/Lazy-Loading)
* [Markdown](https://github.com/BinaryMoon/wp-toolbelt/wiki/Markdown)
* [Monetization](https://github.com/BinaryMoon/wp-toolbelt/wiki/Monetization)
* [Optimization](https://github.com/BinaryMoon/wp-toolbelt/wiki/Optimization)
* [Portfolio](https://github.com/BinaryMoon/wp-toolbelt/wiki/Portfolio)
* [Random Redirection](https://github.com/BinaryMoon/wp-toolbelt/wiki/Random-Redirect)
* [Related Posts](https://github.com/BinaryMoon/wp-toolbelt/wiki/Related-Posts)
* [Responsive Videos](https://github.com/BinaryMoon/wp-toolbelt/wiki/Responsive-Videos)
* [Search Redirect](https://github.com/BinaryMoon/wp-toolbelt/wiki/Search-Redirect)
* [Social Menu](https://github.com/BinaryMoon/wp-toolbelt/wiki/Social-Menu)
* [Spam Blocker](https://github.com/BinaryMoon/wp-toolbelt/wiki/Spam-Blocker)
* [Static Social Sharing](https://github.com/BinaryMoon/wp-toolbelt/wiki/Static-Social-Sharing)
* [Stats](https://github.com/BinaryMoon/wp-toolbelt/wiki/Stats)
* [Testimonials](https://github.com/BinaryMoon/wp-toolbelt/wiki/Testimonials)
* [Typographic Widows](https://github.com/BinaryMoon/wp-toolbelt/wiki/Typographic-Widows)
* [Widget Display](https://github.com/BinaryMoon/wp-toolbelt/wiki/Widget-Display)

## Thanks!

This plugin is heavily inspired by the [Jetpack](https://github.com/automattic/jetpack) and [Machete](https://github.com/nilovelez/machete/) plugins. A lot of the modules use code from these plugins as starting points.


changelog.txt
= 2.7 - 20th August 2020 =
* Add custom avatars as a Gravatar replacement.
* Address CSP issues and re-enable the module.
* Update Infinite Scroll so that it has it's own theme support feature 'toolbelt-infinite-scroll'. Can still use 'infinite-scroll' as used in Jetpack.
* Update Infinite Scroll to focus first loaded article after new page of content loads (improves accessibility).
* Simplify social menu icon css.
* Minor bug fixes.

= 2.6.1 - 27th July 2020 =
* Disable CSP - it's not ready for prime time yet.

= 2.6 - 27th July 2020 =
* Add CSP (Content Security Policy) module.
* Add a breadcrumbs block (for full site editing).
* Add social share settings for an easy way to turn on/ off the different sharing sites.
* Add more social share sites! Instapaper, Flipboard, Tumblr, HackerNews, and Evernote.
* Add 'rel=nofollow' to social sharing links.
* Remove `wp_` prefix from social network filter.

= 2.5.3 - 15th July 2020 =
* Add hover colour to social share links so that they are always readable.
* Ensure the heading anchors don't move the content around.
* Add title to footnotes links.
* Tweak cookie banner css to ensure the box size doesn't change unexpectedly.
* Make columns css more robust so that it will work with every theme.
* Add basic support for the WordPress AMP plugin so that it doesn't output custom CSS on AMP pages.
* Add support for more social sharing sites (Reddit, Pocket, Wallabag, and send by email).
* Add scroll margin top support for heading anchors.

= 2.5.2 - 2nd March 2020 =
* Add new footnotes module that appends footnotes to the bottom of post content.
* Add a search results redirect if there's only a single result.
* Add additional cleanup rules for more speed!
* Only add css properties when a module is active that supports them.
* Automatically update the spam blocker blocklist once a week.

= 2.5.1 - 17th December 2019 =
* Fix infinite scroll button. It now hides when we get to the end of the content.
* Update blocklist.

= 2.5 - 14th December 2019 =
* Added a simple Gist block.
* Refactor widget visibility module. You may need to change your display rules. Check the documentation for more info. https://github.com/BinaryMoon/wp-toolbelt/wiki/Widget-Display
* Add new pagechild and 404 display rules to the widget visibility module.
* Add a contact form 'subject' field which we can use to override the default contact form subject.
* Add a custom post type for saving contact form submissions. The feedback gets deleted regularly so that we don't store user data unnecessarily.
* Add contact form custom subject line modifications.
* Stop sending contact form messages that get flagged as spam. They are still saved to the database and readable through the admin.
* Add fix for typographic widows.
* Add list and blockquote as internal block types for the contact form block.
* Ensure the contact form submit button stays in the right place! :)
* Fix issue with settings link appearing when it shouldn't.
* Fix: Only include contact form validation rules when there is a contact form on the page
* Make GDPR requirements of cookie banner clearer.
* Improve module selector design.
* Make the admin style tweaks show in the customizer (most important for the button improvements).
* Simplify portfolio and testimonial block code a little.
* Rename blacklist to blocklist. Language matters!

= 2.4.0 - 24th November 2019 =
* Add contact form block.
* Make admin button styles clearer (in Admin Tweaks module).
* Update spam blacklist.
* Fix dns prefetch for analytics.
* Ensure scripts and styles do not load before the <html> element.
* Remove emoji prefetch url from header in cleanup.
* Fix bug with removing jquery migrate in cleanup (undefined parameter).
* Dispatch events in cookie banner script. Makes it easier to hook into the events.

= 2.3.2 - 15th November 2019 =
* Tweak admin styles for WP 5.3.
* Add javascript events for accept and decline cookie options on cookie bar.
* Update spam blacklist.

= 2.3.1 - 7th November 2019 =
* Update spam blacklist.
* Fix undefined variable error in spam blocking.
* Ensure projects and testimonials block styles are loaded.
* Prevent scripts from being loaded multiple times.
* Make the featured images on the projects block link to the project.
* Add a method for skipping headings in the heading anchor module. Added support for this to the Projects block.
* Fix translation string. Thanks @pixolin for the patch.

= 2.3 - 24th October 2019 =
* Add a Markdown block.
* Add testimonials post type, shortcode, and block.
* Add projects shortcode and block.
* Add spam blocking support for Gravity Forms.
* Add DNS Prefetching to the stats providers.
* Change social sharing css class name from `toolbelt_social_sharing` to `toolbelt-social-sharing`.
* Change related posts, and social sharing so that they don't appear on password protected pages.
* Fix: stop css styles from trying to be included multiple times.
* Tidy up copy issues.

= 2.2.1 - 7th October 2019 =
* Refactor breadcrumbs again. Slightly longer code, but hopefully clearer what it's doing. And will allow adding caching/ filtering if we want.
* Add spam blocking for Jetpack comment forms (still experimental).

= 2.2 - 4th October 2019 =
* Add new simple comment spam module. It's experimental so tread with care.
* Add md to list of fast404 file types.
* Tweak Menu titles to make things more logical.
* Fix small js bug in cookie banner.
* Improve breadcrumb html.

= 2.1 - 22nd September 2019 =
* Add Admin Tweaks module which has a few small CSS tweaks to the WordPress admin to make it a little easier to use the interface.
* Fix setting link on plugin page.
* Move toolbelt_related_posts filter to make it easier to use.

= 2.0 - 21st September 2019 =
* Move settings pages to a top level menu.
* Add Web Monetization module.
* Add tools link to active modules on module settings page.

index.php
<?php
/**
 * Plugin Name: WP Toolbelt
 * Description: More features, with a focus on privacy and speed.
 * Author: Ben Gillbanks
 * Version: 2.7
 * Author URI: https://prothemedesign.com
 * Text Domain: wp-toolbelt
 *
 * @package toolbelt
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

define( 'TOOLBELT_VERSION', '2.7' );
define( 'TOOLBELT_PATH', plugin_dir_path( __FILE__ ) );
define( 'TOOLBELT_PLUGIN_URL', plugins_url( '', __FILE__ ) );
define( 'TOOLBELT_DIR', basename( TOOLBELT_PATH ) );

if ( ! defined( 'TOOLBELT_DISABLE_ADMIN' ) && is_admin() ) {

	require TOOLBELT_PATH . 'admin/admin.php';
	require TOOLBELT_PATH . 'admin/updates.php';

}


/**
 * Load the enabled modules.
 *
 * @return void
 */
function toolbelt_load_modules() {

	$modules = toolbelt_get_modules();
	$options = toolbelt_get_options();
	$load_css_properties = false;

	foreach ( $modules as $slug => $module ) {

		if ( ! empty( $options[ $slug ] ) && 'on' === $options[ $slug ] ) {

			toolbelt_load_module( $slug, $module );

			if ( isset( $module['supports'] ) && in_array( 'css-properties', $module['supports'], true ) ) {
				$load_css_properties = true;
			}

		}
	}

	/**
	 * If there's at least one moduole active that supports the css custom
	 * properties then enqueue them.
	 */
	if ( $load_css_properties ) {

		add_filter( 'wp_print_styles', 'toolbelt_css_properties' );
		add_filter( 'admin_head', 'toolbelt_css_properties' );

	}

}

toolbelt_load_modules();


/**
 * Load the module and associated admin functionality.
 *
 * @param string        $slug The module slug. Used as the file path.
 * @param array<string> $module The module properties.
 * @return void
 */
function toolbelt_load_module( $slug, $module ) {

	// Load core module code.
	require_once TOOLBELT_PATH . 'modules/' . $slug . '/module.php';

	if ( ! is_admin() ) {

		return;

	}

	if ( ! isset( $module['supports'] ) ) {

		return;

	}

	// Load module tools.
	if ( in_array( 'tools', $module['supports'], true ) ) {

		require_once TOOLBELT_PATH . 'modules/' . $slug . '/tools.php';

	}

	// Load module settings.
	if ( in_array( 'settings', $module['supports'], true ) ) {

		require_once TOOLBELT_PATH . 'modules/' . $slug . '/settings.php';

	}

}


/**
 * Get the list of available Toolbelt modules.
 *
 * @return array<mixed>
 */
function toolbelt_get_modules() {

	return array(
		'admin-tweaks' => array(
			'name' => esc_html__( 'Admin Tweaks', 'wp-toolbelt' ),
			'description' => esc_html__( 'Tweak styles in the admin to make it more usable.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Admin-Tweaks',
		),
		'avatars' => array(
			'name' => esc_html__( 'Avatars', 'wp-toolbelt' ),
			'description' => esc_html__( 'A privacy focused replaced for Gravatar.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Avatars',
			'weight' => esc_html__( '5kb of Javascript, but it removes all Gravatar images so should be an improvement overall', 'wp-toolbelt' ),
			'supports' => array( 'experimental' ),
		),
		'breadcrumbs' => array(
			'name' => esc_html__( 'Breadcrumbs', 'wp-toolbelt' ),
			'description' => esc_html__( 'Simple, fast, breadcrumbs. Requires support from the theme to display. See docs for more info.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Breadcrumbs',
			'weight' => esc_html__( '1 or 2kb of HTML.', 'wp-toolbelt' ),
			'supports' => array( 'css-properties' ),
		),
		'cleanup' => array(
			'name' => esc_html__( 'Header Cleanup', 'wp-toolbelt' ),
			'description' => esc_html__( 'Remove unnecessary HTML from the site header.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Optimization',
			'weight' => esc_html__( 'Minus 5kb of HTML from the page head.', 'wp-toolbelt' ),
		),
		'contact-form' => array(
			'name' => esc_html__( 'Contact Form', 'wp-toolbelt' ),
			'description' => esc_html__( 'Create a contact form.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Contact-Form',
			'supports' => array( 'experimental', 'css-properties' ),
			'weight' => esc_html__( '1.2kb of inline CSS, and 12kb of inline JS.', 'wp-toolbelt' ),
		),
		'cookie-banner' => array(
			'name' => esc_html__( 'Cookie Banner', 'wp-toolbelt' ),
			'description' => esc_html__( 'Display a simple banner with a link to your Privacy Policy.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Cookie-Banner',
			'supports' => array( 'gdpr-hard-mode', 'css-properties' ),
			'weight' => esc_html__( '1.2kb of inline JS and CSS.', 'wp-toolbelt' ),
		),
		'content-security-policy' => array(
			'name' => esc_html__( 'Content Security Policy Headers', 'wp-toolbelt' ),
			'description' => esc_html__( 'Inject a content security policy header into your page responses', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/CSP-Header',
			'supports' => array( 'experimental', 'warning' ),
			'weight' => esc_html__( 'Roughly 1kb of header properties.', 'wp-toolbelt' ),
		),
		'disable-comment-urls' => array(
			'name' => esc_html__( 'Disable the Comment URL Field', 'wp-toolbelt' ),
			'description' => esc_html__( 'Remove the URL field from comment forms. This may only work on the core comment form, and not on custom ones added to themes.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Disable-Comment-Urls',
			'supports' => array( 'tools' ),
		),
		'fast-404' => array(
			'name' => esc_html__( 'Fast 404s', 'wp-toolbelt' ),
			'description' => esc_html__( 'Disable 404 pages for static files like images.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Fast-404',
		),
		'featured-attachment' => array(
			'name' => esc_html__( 'Featured Attachment', 'wp-toolbelt' ),
			'description' => esc_html__( 'If there is no featured image for a post then use the first image attachment instead.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Featured-Attachment',
		),
		'footnotes' => array(
			'name' => esc_html__( 'Footnotes', 'wp-toolbelt' ),
			'description' => esc_html__( 'Add footnotes to the bottom of the page.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Footnotes',
		),
		'gist' => array(
			'name' => esc_html__( 'Gist Embed Block', 'wp-toolbelt' ),
			'description' => esc_html__( 'Easily embed Github Gists onto your site.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Gist',
		),
		'heading-anchors' => array(
			'name' => esc_html__( 'Heading Anchors', 'wp-toolbelt' ),
			'description' => esc_html__( 'Allow site visitors to link to individual sections of the page. Adds unique ids to each heading.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Heading-Anchor',
			'weight' => esc_html__( '0.2kb of inline CSS.', 'wp-toolbelt' ),
		),
		'infinite-scroll' => array(
			'name' => esc_html__( 'Infinite Scroll', 'wp-toolbelt' ),
			'description' => esc_html__( 'Load new post content indefinitely. This may require some changes to your theme for it to work properly.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Infinite-Scroll',
			'weight' => esc_html__( '0.8kb of inline CSS and 1.3kb of inline JS.', 'wp-toolbelt' ),
			'supports' => array( 'css-properties' ),
		),
		'layout-grid' => array(
			'name' => esc_html__( 'Layout Grid', 'wp-toolbelt' ),
			'description' => esc_html__( 'Display content in carefully curated columns.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Layout-Grid',
		),
		'lazy-load' => array(
			'name' => esc_html__( 'Lazy Load images', 'wp-toolbelt' ),
			'description' => esc_html__( 'Add native browser lazy loading to all images on your website. Currently this only works in Chrome, but hopefully it will be added to all browsers.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Lazy-Loading',
		),
		'markdown' => array(
			'name' => esc_html__( 'Markdown', 'wp-toolbelt' ),
			'description' => esc_html__( 'Render Markdown. Uses Github flavor Markdown.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Markdown',
		),
		'monetization' => array(
			'name' => esc_html__( 'Web Monetization', 'wp-toolbelt' ),
			'description' => esc_html__( 'Adds the web monetization meta tag. This allows you to get paid for your content. See docs for more information.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Monetization',
			'supports' => array( 'settings' ),
		),
		'post-listing' => array(
			'name' => esc_html__( 'Post Listing', 'wp-toolbelt' ),
			'description' => esc_html__( 'Display blog posts and other content in a block.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Post-Listing',
		),
		'projects' => array(
			'name' => esc_html__( 'Portfolio', 'wp-toolbelt' ),
			'description' => esc_html__( 'A portfolio custom post type for your projects.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Portfolio',
			'supports' => array( 'tools', 'css-properties' ),
		),
		'random-redirect' => array(
			'name' => esc_html__( 'Random Redirect', 'wp-toolbelt' ),
			'description' => esc_html__( 'Randomly Redirect to a blog post.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Random-Redirect',
		),
		'related-posts' => array(
			'name' => esc_html__( 'Related Posts', 'wp-toolbelt' ),
			'description' => esc_html__( 'Speedy related posts.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Related-Posts',
			'weight' => esc_html__( '0.3kb of inline CSS, plus the HTML and images.', 'wp-toolbelt' ),
			'supports' => array( 'tools', 'css-properties' ),
		),
		'responsive-videos' => array(
			'name' => esc_html__( 'Responsive Videos', 'wp-toolbelt' ),
			'description' => esc_html__( 'Ensure embedded videos maintain a 16:9 aspect ratio on all screen sizes. Ignores blocks with responsive videos.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Responsive-Videos',
			'weight' => esc_html__( '0.2kb of inline CSS.', 'wp-toolbelt' ),
		),
		'search-redirect' => array(
			'name' => esc_html__( 'Search Redirect', 'wp-toolbelt' ),
			'description' => esc_html__( 'If there is a single search result then redirect to the result, rather than displaying a list of results.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Search-Redirect',
		),
		'social-menu' => array(
			'name' => esc_html__( 'Social Menu', 'wp-toolbelt' ),
			'description' => esc_html__( 'Add a social icons menu. This must be integrated into the theme.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Social-Menu',
			'weight' => esc_html__( '0.2kb of inline CSS, plus the SVGs needed for the icons.', 'wp-toolbelt' ),
		),
		'social-sharing' => array(
			'name' => esc_html__( 'Static Social Sharing', 'wp-toolbelt' ),
			'description' => esc_html__( 'Add social sharing links that use the platforms native sharing system.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Static-Social-Sharing',
			'weight' => esc_html__( '4.1kb of inline SVG icons, and 0.7kb of inline CSS.', 'wp-toolbelt' ),
			'supports' => array( 'css-properties', 'settings' ),
		),
		'spam-blocker' => array(
			'name' => esc_html__( 'Spam Blocker', 'wp-toolbelt' ),
			'description' => esc_html__( 'Attempt to block spam.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Spam-Blocker',
			'weight' => esc_html__( '0.3kb of inline JS', 'wp-toolbelt' ),
			'supports' => array( 'experimental' ),
		),
		'star-rating' => array(
			'name' => esc_html__( 'Star Ratings', 'wp-toolbelt' ),
			'description' => esc_html__( 'Rate movies, books, videos, difficulty, whatever.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Star-Ratings',
		),
		'stats' => array(
			'name' => esc_html__( 'Stats', 'wp-toolbelt' ),
			'description' => esc_html__( 'Website statistics and analytics.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Stats',
			'supports' => array( 'settings' ),
		),
		'testimonials' => array(
			'name' => esc_html__( 'Testimonial', 'wp-toolbelt' ),
			'description' => esc_html__( 'A testimonials custom post type.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Testimonials',
			'weight' => esc_html__( '0.5kb of inline CSS.', 'wp-toolbelt' ),
			'supports' => array( 'tools', 'css-properties' ),
		),
		'widows' => array(
			'name' => esc_html__( 'Typographic Widows', 'wp-toolbelt' ),
			'description' => esc_html__( 'Fix typographic widows in titles and headings.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Typographic-Widows',
		),
		'widget-display' => array(
			'name' => esc_html__( 'Widget Display', 'wp-toolbelt' ),
			'description' => esc_html__( 'Control which pages widgets appear on.', 'wp-toolbelt' ),
			'docs' => 'https://github.com/BinaryMoon/wp-toolbelt/wiki/Widget-Display',
		),
	);

}


/**
 * Output custom css properties that are used by the plugin css.
 *
 * This can be used by themes to make the spacings match those used.
 *
 * @return void
 */
function toolbelt_css_properties() {

	if ( toolbelt_is_amp_page() ) {
		return;
	}

	$properties = array(
		'toolbelt-spacing' => '1rem',
		'toolbelt-scroll-margin-top' => '45px',
	);

	$properties = apply_filters(
		'toolbelt_css_properties',
		$properties
	);

	$css_properties = '';

	foreach ( $properties as $key => $value ) {
		$css_properties .= ' --' . $key . ':' . $value . ';';
	}

	echo '<style name="toolbelt-properties">:root {';
	echo esc_attr( $css_properties );
	echo '}</style>';

}


/**
 * Inline the module css.
 *
 * @param string $module The module slug.
 * @param string $file The file name for the style to include.
 * @return void
 */
function toolbelt_styles( $module, $file = 'style' ) {

	if ( defined( 'TOOLBELT_DISABLE_STYLES' ) ) {
		return;
	}

	if ( toolbelt_is_amp_page() ) {
		return;
	}

	$css_filter = sprintf( 'toolbelt_hide_%s_styles', $module );

	if ( apply_filters( $css_filter, false ) ) {
		return;
	}

	$path = TOOLBELT_PATH . 'modules/' . $module . '/' . $file . '.min.css';

	if ( in_array( $path, get_included_files(), true ) ) {
		return;
	}

	echo '<style name="toolbelt-style-' . esc_attr( $module ) . '">';
	require_once $path;
	echo '</style>';

}


/**
 * Load the block editor styles.
 *
 * @param string $module The module to load.
 * @param string $file The file to load. Defaults to block.css.
 * @return void|null
 */
function toolbelt_styles_editor( $module, $file = 'block' ) {

	$screen = get_current_screen();

	if ( ! $screen || 'post' !== $screen->base ) {
		return;
	}

	toolbelt_styles( $module, $file );

}


/**
 * Inline the module css.
 *
 * @param string $module The module slug.
 * @return void|null
 */
function toolbelt_global_styles( $module ) {

	if ( defined( 'TOOLBELT_DISABLE_STYLES' ) ) {
		return;
	}

	$css_filter = sprintf( 'toolbelt_hide_%s_styles', $module );

	if ( apply_filters( $css_filter, false ) ) {
		return;
	}

	$path = TOOLBELT_PATH . 'assets/css/' . $module . '.css';

	if ( in_array( $path, get_included_files(), true ) ) {
		return;
	}

	echo '<style name="toolbelt-global-style-' . esc_attr( $module ) . '">';
	require_once $path;
	echo '</style>';

}


/**
 * Inline the module script.
 *
 * @param string $module The module slug.
 * @param string $file Optional filename to include. Defaults to 'script'.
 * @return void|null
 */
function toolbelt_scripts( $module, $file = 'script' ) {

	// Output scripts.
	$path = TOOLBELT_PATH . 'modules/' . $module . '/' . $file . '.min.js';

	if ( in_array( $path, get_included_files(), true ) ) {
		return;
	}

	echo '<script name="toolbelt-script-' . esc_attr( $module ) . '">';
	require_once $path;
	echo '</script>';

}


/**
 * Inline the module script.
 *
 * @param string $script The module slug.
 * @return void|null
 */
function toolbelt_global_script( $script ) {

	// Output scripts.
	$path = TOOLBELT_PATH . 'assets/js/' . $script . '.js';

	if ( in_array( $path, get_included_files(), true ) ) {
		return;
	}

	echo '<script name="toolbelt-script-' . esc_attr( sanitize_title( $script ) ) . '">';
	require_once $path;
	echo '</script>';

}


/**
 * Load the toolbelt options.
 *
 * @return array<string>
 */
function toolbelt_get_options() {

	/**
	 * Add a filter that allows us to override the database options.
	 *
	 * If these options are set then they will be the only options used. The
	 * database will be ignored entirely.
	 *
	 * @see https://github.com/BinaryMoon/wp-toolbelt/wiki/Admin-Settings#toolbelt_options
	 */
	$options = apply_filters( 'toolbelt_options', null );

	if ( empty( $options ) ) {

		$options = get_option( 'toolbelt_options' );

	}

	return $options;

}


/**
 * Display the plugin url relative to the plugin root.
 *
 * @param string $path The path to load.
 * @return string
 */
function toolbelt_plugins_url( $path ) {

	return TOOLBELT_PLUGIN_URL . '/' . ltrim( $path, '/' );

}


/**
 * Add a block category for Toolbelt.
 *
 * @param array<mixed> $categories The current list of categories.
 * @param string       $post The post type for the current page.
 * @return array<mixed>
 */
function toolbelt_block_category( $categories, $post ) {

	return array_merge(
		$categories,
		array(
			array(
				'slug' => 'wp-toolbelt',
				'title' => esc_html__( 'Toolbelt', 'toolbelt' ),
			),
		)
	);

}


/**
 * Register block categories.
 *
 * Call this directly from modules. It will only add the filter once.
 *
 * @return void
 */
function toolbelt_register_block_category() {

	if ( has_filter( 'block_categories', 'toolbelt_block_category' ) ) {
		return;
	}

	add_filter( 'block_categories', 'toolbelt_block_category', 10, 2 );

}


/**
 * If is AMP page.
 *
 * @return boolean
 */
function toolbelt_is_amp_page() {

	if ( function_exists( 'is_amp_endpoint' ) && is_amp_endpoint() ) {
		return true;
	}

	return false;

}
